StatisticsV2: statistics framework initial redesign for Datafusion

Fly-Style opened the pull request:
## Rationale for this change
https://synnada.notion.site/Redesigning-and-Enhancing-the-Statistics-Framework-in-Datafusion-16bf46d2dab180448272dbbd1d1f7cea

## What changes are included in this PR?

This patch presents a Statistics v.2 framework with the following main points:
- introduction to enum-based struct to support multiple distribution types, which initially include:
  - Uniform distribution (range)
  - Gaussian distribution (also known as normal)
  - Exponential distribution
  - Bernoulli distribution (holds probability, is used as resulting distribution of comparison operators),
  - Unknown distribution, abstracts any non-represented distribution.
- tree-based stats calculation is divided into bottom-up evaluation and top-down propagation;

### Tables of statistic execution and propagation rules for `PhysicalExpr`-s:

#### Definitions
<details>
UF = Uniform
UN = Unknown
EXP = Exponential
GSS = Gaussian
BRN = Bernoulli
</details>

#### Binary arithmetical operators, evaluation.
<details>

| Input 1           | Input 2           | (+)             | (-) | (*) | (/) |
|-------------------|-------------------|-----------------|-----|-----|-----|
| **UF** [a,b]      | **UF** [c,d]      | UN              | UN  | UN  | UN  |
| **UF** [a,b]      | **GSS** N(μ,σ²)   | UN              | UN  | UN  | UN  |
| **UF** [a,b]      | **EXP** Exp(λ)    | UN              | UN  | UN  | UN  |
| **UF** [a,b]      | **BRN**           | UN              | UN  | UN  | UN  |
| **UF** [a,b]      | **UN**            | UN              | UN  | UN  | UN  |
| **GSS** N(μ₁,σ₁²) | **GSS** N(μ₂,σ₂²) | GSS             | GSS | GSS | UN  |
| **GSS** N(μ,σ²)   | **EXP** Exp(λ)    | UN              | UN  | UN  | UN  |
| **GSS** N(μ,σ²)   | **BRN**           | UN              | UN  | UN  | UN  |
| **GSS** N(μ,σ²)   | **UN**            | UN              | UN  | UN  | UN  |
| **EXP** Exp(λ₁)   | **EXP** Exp(λ₂)   | UN (gamma dist) | UN  | UN  | UN  |
| **EXP** Exp(λ)    | **BRN**           | UN              | UN  | UN  | UN  |                                   
| **EXP** Exp(λ)    | **UN**            | UN              | UN  | UN  | UN  |                                   
| **BRN** Exp(λ)    | **BRN**           | UN              | UN  | UN  | UN  |                                   
| **BRN** Exp(λ)    | **UN**            | UN              | UN  | UN  | UN  |                                   
| **UN**            | **UN**            | UN              | UN  | UN  | UN  |
</details>

#### Comparison operators, evaluation.

<details>

| Input 1           | Input 2           | Comparison (=, ≠) | Inequality (>, >=) | Inequality (<, <=) |
|-------------------|-------------------|-------------------|--------------------|--------------------|
| **UF** [a,b]      | **UF** [c,d]      | BRN               | BRN                | BRN                |
| **GSS** N(μ₁,σ₁²) | **GSS** N(μ₂,σ₂²) | UN                | UN(Ф-func)         | UN(Ф-func)         |
| **EXP** Exp(λ₁)   | **EXP** Exp(λ₂)   | UN                | UN                 | UN                 |
| **BRN**           | **BRN**           | BRN               | BRN                | BRN                |
| **UF** [a,b]      | **GSS** N(μ,σ²)   | UN                | UN                 | UN                 |
| **UF** [a,b]      | **EXP** Exp(λ)    | UN                | UN                 | UN                 |
| **UF** [a,b]      | **BRN**           | UN?               | UN?                | UN?                |
| **UF** [a,b]      | **UN**            | BRN?              | BRN?               | BRN?               |
| **GSS** N(μ,σ²)   | **EXP** Exp(λ)    | UN                | UN                 | UN                 |
| **GSS** N(μ,σ²)   | **BRN**           | BRN?/UN           | BRN?/UN            | BRN? /UN           |
| **GSS** N(μ,σ²)   | **UN**            | UN                | UN                 | UN                 |
| **EXP** Exp(λ)    | **UN**            | UN, estimate      | UN, estimate       | UN, estimate       |                                   
| **EXP** Exp(λ)    | **BRN**           | UN, estimate      | UN, estimate       | UN, estimate       |                                   
| **UN**            | **BRN**           | BRN/UN            | BRN/UN             | BRN/UN             |
| **UN**            | **UN**            | BRN/UN            | BRN/UN             | BRN/UN             |

</details>

## Are these changes tested?

These changes were massively tested, you can find a lot of unit tests coupled to this feature.

Commit:
StatisticsV2: initial definition and validation method implementation

Fly-Style committed:
Implement mean, median and standard deviation extraction for StatsV2

Fly-Style committed:
Move stats_v2 to `physical-expr` package

Fly-Style committed:
Introduce `ExprStatisticGraph` and `ExprStatisticGraphNode`

Fly-Style committed:
Split the StatisticsV2 and statistics graph locations, prepare the infrastructure for stats top-down propagation and final bottom-up calculation

Fly-Style committed:
Merge branch 'apache_main' into feat/statistics_v2

Fly-Style committed:
Create a skeleton for statistics bottom-up evaluation

Fly-Style committed:
Calculate variance instead of std_dev

Fly-Style committed:
Introduce high-level test for 'evaluate_statistics()'

Fly-Style committed:
Refactor result distribution computation during the statistics evaluation phase; add compute_range function

Fly-Style committed:
Always produce Unknown distribution in non-mentioned combination cases, todos for the future

Fly-Style committed:
Introduce Bernoulli distribution to be used as result of comparisons and inequations distribution combinations

Fly-Style committed:
Implement initial statistics propagation of Uniform and Unknown distributions with known ranges

Fly-Style committed:
Implement evaluate_statistics for logical not and unary negation operator

berkaysynnada commented on datafusion/physical-expr-common/src/physical_expr.rs:
```diff
@@ -144,6 +145,26 @@ pub trait PhysicalExpr: Send + Sync + Display + Debug + DynEq + DynHash {
+        _parent_stat: &StatisticsV2,
+        _children_stat: &[&StatisticsV2],
+    ) -> Result<Option<Vec<StatisticsV2>>> {
+        Ok(Some(vec![]))
```
I know this is the logic that used in propagate_constraints --returning an empty vector instead of cloning `children` if none of the children intervals change as a result of propagation-- should we leave this behavior for simplicity and expressiveness? I am also thinking that changing the propagate_constraints in this manner. I think this hack is bringing more debug or code complexity cost than its non-clone efficiency

It's not a big deal but I'd like to discuss it.

Fly-Style replied:
> returning an empty vector instead of cloning children if none of the children intervals change as a result of propagation

The original option with empty vec is safer from OOP point of view. The complete solution should be implemented like you've mentioned, but **all** expressions should be revisited.

berkaysynnada commented on datafusion/physical-expr-common/src/stats.rs:
```diff
@@ -0,0 +1,738 @@
+}
+
+impl StatisticsV2 {
+    pub fn new_unknown() -> Self {
```
If we can force all users to instantiate a statistics object with a known datatype, then we can start with a [-inf, +inf] range with that datatype. Mutating the datatype of the range in runtime is like a design smell to me. WDYT?

Fly-Style replied:
Totally agree.

berkaysynnada commented on datafusion/physical-expr-common/src/stats.rs:
```diff
@@ -0,0 +1,738 @@
+    /// - [`Gaussian`] distribution has it explicitly
+    /// - [`Bernoulli`] mean is `p`
+    /// - [`Unknown`] distribution _may_ have it explicitly
+    pub fn mean(&self) -> datafusion_common::Result<Option<ScalarValue>> {
```
If not valid, then error,
If unknown, then Ok(None() maybe?

berkaysynnada commented on datafusion/physical-expr-common/src/stats.rs:
```diff
@@ -0,0 +1,738 @@
+    /// - [`Uniform`]'s range is its interval
+    /// - [`Bernoulli`]'s range is always [0, 1], but we return [`Interval::UNCERTAIN`]
+    /// - [`Unknown`]'s range is unbounded by default, but
+    pub fn range(&self) -> Option<&Interval> {
```
We can add gaussian and exponential ranges maybe? with infinite bounds

Fly-Style replied:
It may be worth it, but, unfortunately, we can't infer the correct data type. The only foreseeable option is using `DataType:Null`, but not sure definitely.

berkaysynnada replied:
Can we require from user giving a datatype when initializing any distribution?

berkaysynnada reviewed (commented):
I really like how the code is organized, and your clean coding style, even though it’s still draft.

While reviewing `ExprStatisticGraph`, I wondered if introducing another distribution, **singleton**, to represent constants and known values, might improve us. It seems like this could simplify the representation of literals significantly. However, I’m not sure if it would complicate the computation process. Let’s discuss this too.

I’ll also think more about the propagation of distributions. However, it seems unlikely that the current approach for range propagations can be improved further.

Fly-Style committed:
Fix and add tests; make fmt happy

Fly-Style committed:
Add integration test, implement conversion into Bernoulli distribution for Eq and NotEq

Fly-Style committed:
Finish test, small cleanup

berkaysynnada commented on datafusion/physical-expr-common/src/stats.rs:
```diff
@@ -0,0 +1,762 @@
+    /// - [`Bernoulli`] mean is `p`
+    /// - [`Unknown`] distribution _may_ have it explicitly
+    pub fn mean(&self) -> datafusion_common::Result<Option<ScalarValue>> {
+        if !self.is_valid() {
```
Users can construct an invalid Statistics, but they realize it once they access to its mean/median etc. I wonder if it would be better if we disable construction by the enum itself directly, but implement the constructor utils like StatisticsV2::new_uniform(mean: Option<ScalarValue>, ...), and make these validation checks there. WDYT?

berkaysynnada replied:
So, we just need to check while initializing the objects, but nowhere else. Otherwise, we always need to be careful and add that check before each knowledge extraction or mathematical operation.

Fly-Style replied:
That's a wise proposition, gonna check stats validity on creation and implement ctors.

Fly-Style replied:
https://github.com/synnada-ai/datafusion-upstream/pull/57/commits/b8068b86f03ef63fc929302d785aaa03483f2419

berkaysynnada commented on datafusion/physical-expr-common/src/stats.rs:
```diff
@@ -0,0 +1,762 @@
+    /// Extract the range of given statistic distribution:
+    /// - [`Uniform`]'s range is its interval
+    /// - [`Bernoulli`]'s range is always [0, 1], but we return [`Interval::UNCERTAIN`]
+    /// - [`Unknown`]'s range is unbounded by default, but
```
Exponential's range will be [offset, infinite], and also when p is 0 or 1, we can say CERTAINLY_FALSE or _TRUE for Bernoulli. Can you add those as well in both docs and method's itself?

Fly-Style replied:
https://github.com/synnada-ai/datafusion-upstream/pull/57/commits/b8068b86f03ef63fc929302d785aaa03483f2419

berkaysynnada commented on datafusion/physical-expr-common/src/stats.rs:
```diff
@@ -0,0 +1,762 @@
+        match &self {
+            Uniform { interval, .. } => Some(interval.data_type()),
+            Bernoulli { p } => Some(p.data_type()),
+            Unknown { range, .. } => Some(range.data_type()),
```
Offset of the exponential can be the datatype of Exponential Distribution?

Fly-Style replied:
Sure!

Fly-Style replied:
https://github.com/synnada-ai/datafusion-upstream/pull/57/commits/b8068b86f03ef63fc929302d785aaa03483f2419

berkaysynnada commented on datafusion/physical-expr-common/src/stats.rs:
```diff
@@ -0,0 +1,762 @@
+    /// f(x, λ) = (λe)^-λx, if x >= 0
+    Exponential {
+        rate: ScalarValue,
+        offset: ScalarValue,
```
I guess we might have a conflict for the meaning of offset. Perhaps using the naming "offset" was problematic -- I should have used "shift", and it updates the formula such:
`f(x, λ, offset) = (λe)^-λ(x-offset), if x >= offset`

Did you think it like
`f(x, λ, offset) = offset * (λe)^-λ(x), if x >= 0`

berkaysynnada replied:
Let's be clear on this, as my suggestions for range and datatype rely on this assumption.

Fly-Style replied:
@berkaysynnada ah, roger that, will update it!

Fly-Style replied:
https://github.com/synnada-ai/datafusion-upstream/pull/57/commits/b8068b86f03ef63fc929302d785aaa03483f2419

berkaysynnada commented on datafusion/physical-expr/src/expressions/binary.rs:
```diff
@@ -494,6 +501,133 @@ impl PhysicalExpr for BinaryExpr {
+    ) -> Result<StatisticsV2> {
+        let (left_stat, right_stat) = (children_stat[0], children_stat[1]);
+
+        // We can evaluate statistics only with numeric data types on stats.
```
I think we can remove these type checks since operations' themself do also those checks. Am I missing a point?

Fly-Style replied:
https://github.com/synnada-ai/datafusion-upstream/pull/57/commits/b8068b86f03ef63fc929302d785aaa03483f2419

berkaysynnada commented on datafusion/physical-expr/src/expressions/binary.rs:
```diff
@@ -494,6 +501,133 @@ impl PhysicalExpr for BinaryExpr {
+        match &self.op {
+            Operator::Plus | Operator::Minus | Operator::Multiply => {
+                match (left_stat, right_stat) {
+                    (
```
To keep this part more clear, let's handle this gaussian case in a separate function?

Fly-Style replied:
https://github.com/synnada-ai/datafusion-upstream/pull/57/commits/b8068b86f03ef63fc929302d785aaa03483f2419

berkaysynnada commented on datafusion/physical-expr/src/expressions/binary.rs:
```diff
@@ -494,6 +501,133 @@ impl PhysicalExpr for BinaryExpr {
+                    }
+                    // TODO: complement with more cases
+                    (_, _) => {
+                        new_unknown_from_binary_expr(&self.op, left_stat, right_stat)
```
Let's wrap this `And` and `Or` blocks with a function as well

Fly-Style replied:
https://github.com/synnada-ai/datafusion-upstream/pull/57/commits/b8068b86f03ef63fc929302d785aaa03483f2419

berkaysynnada commented on datafusion/physical-expr/src/expressions/negative.rs:
```diff
@@ -140,6 +145,85 @@ impl PhysicalExpr for NegativeExpr {
+    fn evaluate_statistics(&self, stats: &[&StatisticsV2]) -> Result<StatisticsV2> {
+        assert_eq!(stats.len(), 1);
+
+        if !stats[0].is_valid() {
```
All of these checks will be redundant once we check during initialization

Fly-Style replied:
Removed

Fly-Style replied:
https://github.com/synnada-ai/datafusion-upstream/pull/57/commits/b8068b86f03ef63fc929302d785aaa03483f2419

berkaysynnada committed:
minor improvements

berkaysynnada reviewed (commented):
I've sent a commit that has mostly minor code improvements. If you see any arguable change, we can discuss it.

What I see as a further improvement is implementing CastExpr methods as it is used very widely. Another major part is for ScalarFunctionExpr's. If we can also practice one of them (we can select the most trivial one), that would be highly beneficial.

berkaysynnada commented:
Once we resolve these last items, @ozankabak will take a final look, and we can upstream this PR.

berkaysynnada committed:
Update stats.rs

berkaysynnada commented on datafusion/physical-expr/src/utils/stats.rs:
```diff
@@ -0,0 +1,687 @@
+/// Computes a median value for a given binary operator and two statistics.
+/// The median is calculable only between:
+/// [`Uniform`] and [`Uniform`] distributions,
+/// [`Gaussian`] and [`Gaussian`] distributions,
```
I guess you've forgotten the gaussian case here. Once you add it, the clippy complains no more

berkaysynnada replied:
My commit changed the expected results -- as it returns error now for invalid types rather than false. I've done so because IMO we should distinguish invalid cases -- whether it is invalid because of argument is not properly set wrt. mathematical rules (user error), or there is a datatype or interval arithmetic related error (which is an internal error actually).

Fly-Style replied:
> I guess you've forgotten the gaussian case here. Once you add it, the clippy complains no more

Yes, I will add in the consecutive commit, a bit later.

Fly-Style replied:
Oh, I didn't notice `rstest` availability, woulda use it, but have no will to rewrite tests now.

Fly-Style replied:
https://github.com/synnada-ai/datafusion-upstream/pull/57/commits/140fb5e17955aa39a6243f129391e12e3e44c7b3

Fly-Style committed:
Addressing review comments

Fly-Style committed:
Implement median colmputation for Gaussian-Gaussian pair

berkaysynnada commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,734 @@
+    Uniform {
+        interval: Interval,
+    },
+    /// f(x, λ, offset) = (λe)^-λx, if x >= offset
```
```suggestion
    /// f(x, λ, offset) = (λe)^(-λ(x - offset)), if x >= offset
```

Fly-Style replied:
https://github.com/synnada-ai/datafusion-upstream/pull/57/commits/f4dd402f8a07121a9e6c411a09b15758cb405725

berkaysynnada commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,734 @@
+    /// and checks newly created statistic on validness.
+    pub fn new_uniform(interval: Interval) -> Result<Self> {
+        let stat = Uniform { interval };
+        if stat.is_valid()? {
```
This check does nothing for uniform. I don't know if you kept it to maintain consistency. If that’s the case, I agree, as additional checks for uniform can be added here in the future, and seeing a similar pattern seems clear.

Fly-Style replied:
> I don't know if you kept it to maintain consistency

Correct.

berkaysynnada committed:
Update stats_v2.rs

berkaysynnada commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,734 @@
+                }
+
+                if range.lower().is_null() && range.upper().is_null()
+                    || !range.lower().is_null() && !range.upper().is_null()
```
`is_null()` doesn't just check whether a value is ScalarValue::Null; it also returns true for None values of other data types. Because of this, we cannot use create semi-infinite ranges here.

What we actually need to avoid are intervals like [ScalarValue::Null, ScalarValue::UInt64(None)], but that is already enforced during interval creation. 
TLDR; No additional checks are needed here, IMO.

As a side note, my ultimate goal for Interval objects is to eliminate the use of ScalarValue::Null entirely. To achieve this, we need to support interval arithmetic for all comparable data types.

Fly-Style replied:
Okay, will remove these checks. Thanks for pointing that!

Fly-Style replied:
https://github.com/synnada-ai/datafusion-upstream/pull/57/commits/f4dd402f8a07121a9e6c411a09b15758cb405725

berkaysynnada commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,734 @@
+    pub fn data_type(&self) -> Option<DataType> {
+        match &self {
+            Uniform { interval, .. } => Some(interval.data_type()),
+            Bernoulli { p } => Some(p.data_type()),
```
I am a bit confused about Bernoulli. Should we return boolean or `p.data_type()` as you did 🤔

Fly-Style replied:
> Bernoulli Distribution is a type of discrete probability distribution where every experiment conducted asks a question that can be answered only in yes or no.

It should be `Boolean`, good catch!

Fly-Style replied:
https://github.com/synnada-ai/datafusion-upstream/pull/57/commits/f4dd402f8a07121a9e6c411a09b15758cb405725

berkaysynnada commented on datafusion/physical-expr/src/utils/stats_v2_graph.rs:
```diff
@@ -0,0 +1,699 @@
+use std::sync::Arc;
```
Can you add the apache license header?

Fly-Style replied:
https://github.com/synnada-ai/datafusion-upstream/pull/57/commits/f4dd402f8a07121a9e6c411a09b15758cb405725

berkaysynnada commented on datafusion/physical-expr/src/utils/stats_v2_graph.rs:
```diff
@@ -0,0 +1,699 @@
+    /// Creates a DAEG node from DataFusion's [`ExprTreeNode`] object. Literals are creating
+    /// [`Uniform`] distribution kind of statistic with definite, singleton intervals.
+    /// Otherwise, create [`Unknown`] statistic with an unbounded interval.
+    pub fn make_node_with_stats(
```
This API is not used now, but it might be useful. I prefer keeping it, but I guess you forget to change its docstring (because it's same of make_node now)

Fly-Style replied:
You caught it right -- it is not used _yet_. Fixed docs.

Fly-Style replied:
https://github.com/synnada-ai/datafusion-upstream/pull/57/commits/f4dd402f8a07121a9e6c411a09b15758cb405725

berkaysynnada commented on datafusion/physical-expr/src/utils/stats_v2_graph.rs:
```diff
@@ -0,0 +1,699 @@
+
+    pub fn statistic(&self) -> &StatisticsV2 {
+        &self.statistics
+    }
```
maybe adding
```
pub fn expression(&self) -> Arc<dyn PhysicalExpr>{
         Arc::clone(&self.expr)
}
```
as well?

Fly-Style replied:
https://github.com/synnada-ai/datafusion-upstream/pull/57/commits/f4dd402f8a07121a9e6c411a09b15758cb405725

berkaysynnada commented on datafusion/physical-expr/src/utils/stats_v2_graph.rs:
```diff
@@ -0,0 +1,699 @@
+    /// Returns false, if propagation was infeasible, true otherwise.
+    pub fn propagate(&mut self) -> Result<bool> {
+        let mut bfs = Bfs::new(&self.graph, self.root);
+
```
I think you need to check the given range for propagation, like `propagate_constraints()` of `ExprIntervalGraph` does, and this API also should return `PropagationResult` to get more meaning from the propagation result.

berkaysynnada commented on datafusion/physical-expr/src/utils/stats_v2_graph.rs:
```diff
@@ -0,0 +1,699 @@
+
+            // Note: all distributions are recognized as independent, by default.
+            if !children_statistics.is_empty() {
+                self.graph[idx].statistics = self.graph[idx]
```
There is a reverse() call there in the interval evaluation part. Isn't it a problem not having it here?

Fly-Style replied:
Well, I don't see a problem here, it follows the classical BFS. Am I missing something?

berkaysynnada commented on datafusion/physical-expr-common/src/physical_expr.rs:
```diff
@@ -144,6 +145,26 @@ pub trait PhysicalExpr: Send + Sync + Display + Debug + DynEq + DynHash {
+    /// Finds the final statistic of the expression by combining the input statistics
+    /// in a post-order bottom-up manner (post-order DFS in statistics graph).
+    ///
+    /// The part of [`StatisticsV2`] framework, work in progress.
```
Should we remove these "work in progress" statements now?

Fly-Style replied:
https://github.com/synnada-ai/datafusion-upstream/pull/57/commits/f4dd402f8a07121a9e6c411a09b15758cb405725

berkaysynnada commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,734 @@
+use crate::stats_v2::StatisticsV2::{Bernoulli, Exponential, Gaussian, Uniform, Unknown};
```
Let's add apache license to this new file

Fly-Style replied:
https://github.com/synnada-ai/datafusion-upstream/pull/57/commits/f4dd402f8a07121a9e6c411a09b15758cb405725

berkaysynnada commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,734 @@
+        }
+    }
+
+    pub fn new_unknown_with_uncertain_range() -> Self {
```
WDYT renaming this as `new_unknown_boolean()`?

Fly-Style replied:
Eh, being honest, doesn't resonate with me.

berkaysynnada commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,734 @@
+        } else {
+            internal_err!("Tried to construct invalid Uniform statistic")
+        }
+    }
```
We will require a util that takes a ScalarValue only, and construct a statistics of it as it is a singleton value. We can add it now. You've already needed that actually in `make_node()` of `ExprStatisticGraphNode`

berkaysynnada commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,734 @@
+                agg.div(ScalarValue::Float64(Some(2.))).map(Some)
+            }
+            Exponential { rate, .. } => {
+                // Note: better to move it to constant with lazy_static! { ... },
```
Do you want to try oncelock?
```rust
static LN_TWO: OnceLock<ScalarValue> = OnceLock::new();

fn get_ln_two() -> &'static ScalarValue {
    LN_TWO.get_or_init(|| ScalarValue::Float64(Some(2_f64.ln())))
}

Exponential { rate, .. } => {
    get_ln_two().div(rate).map(Some)
}
```

Fly-Style replied:
That's interesting, thanks!

Fly-Style replied:
https://github.com/synnada-ai/datafusion-upstream/pull/57/commits/f4dd402f8a07121a9e6c411a09b15758cb405725

berkaysynnada commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,734 @@
+    /// Extract the median value of given statistic, available for given statistic kinds:
+    /// - [`Uniform`] distribution median is equals to mean: (a + b) / 2
+    /// - [`Exponential`] distribution median is calculable by formula: ln2/λ. λ must be non-negative.
+    /// - [`Gaussian`] distribution median is equals to mean, which is present explicitly.
```
Let's add Bernoulli to be consistent

Fly-Style replied:
https://github.com/synnada-ai/datafusion-upstream/pull/57/commits/f4dd402f8a07121a9e6c411a09b15758cb405725

berkaysynnada commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,734 @@
+    /// Extract the variance of given statistic distribution:
+    /// - [`Uniform`]'s variance is equal to (upper-lower)^2 / 12
+    /// - [`Exponential`]'s variance is equal to mean value and calculable as 1/λ^2
+    /// - [`Gaussian`]'s variance is available explicitly
```
Let's add Bernoulli here too

Fly-Style replied:
https://github.com/synnada-ai/datafusion-upstream/pull/57/commits/f4dd402f8a07121a9e6c411a09b15758cb405725

berkaysynnada commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,734 @@
+    /// - [`Exponential`]'s range is [offset, +inf)
+    /// - [`Unknown`]'s range is unbounded by default, but
+    pub fn range(&self) -> Result<Option<Interval>> {
+        match &self {
```
I am thinking adding Gaussian as [-inf, + inf] here, but with which datatype I am not sure. Perhaps by looking its mean? WDYT

Fly-Style replied:
Well, I don't have strong opinion here, but if we want to have it, let it be `Float64` (we're operating on non-discrete space).

Fly-Style replied:
https://github.com/synnada-ai/datafusion-upstream/pull/57

Fly-Style replied:
https://github.com/synnada-ai/datafusion-upstream/pull/57/commits/f4dd402f8a07121a9e6c411a09b15758cb405725

berkaysynnada committed:
minor improvements

berkaysynnada reviewed (commented):
I've left some suggestions and there are a few points to discuss

Fly-Style committed:
Addressing second review comments, part 1

Fly-Style committed:
Return true in other cases

Fly-Style committed:
Finish addressing review requrests, part 2

berkaysynnada committed:
final clean-up

berkaysynnada committed:
final clean-up

ozankabak commented on datafusion/physical-expr-common/src/physical_expr.rs:
```diff
@@ -144,6 +145,27 @@ pub trait PhysicalExpr: Send + Sync + Display + Debug + DynEq + DynHash {
     }
 
+    /// Finds the final statistic of the expression by combining the input statistics
+    /// in a post-order bottom-up manner (post-order DFS in statistics graph).
```
```suggestion
    /// Computes the final statistic of the expression by combining the input statistics
    /// via a post-order, bottom-up traversal of the statistics graph.
```

ozankabak commented on datafusion/physical-expr-common/src/physical_expr.rs:
```diff
@@ -144,6 +145,27 @@ pub trait PhysicalExpr: Send + Sync + Display + Debug + DynEq + DynHash {
+    /// If statistics propagation reveals an infeasibility for any child, returns
+    /// [`None`]. If none of the children intervals change as a result of propagation,
+    /// may return an empty vector instead of cloning `children`. This is the default
+    /// (and conservative) return value.
```
```suggestion
    /// Updates children statistics using the given parent statistic for this expression.
    /// This is used to propagate constraints down through an expression tree.
    ///
    /// If statistics propagation reveals an infeasibility for any child, returns
    /// [`None`]. If none of the children statistics change as a result of propagation,
    /// may return an empty vector instead of cloning `children`. This is the default
    /// (and conservative) return value.
```

ozankabak commented on datafusion/physical-expr/src/expressions/negative.rs:
```diff
@@ -28,13 +28,17 @@ use arrow::{
+use datafusion_physical_expr_common::stats_v2::StatisticsV2;
+use datafusion_physical_expr_common::stats_v2::StatisticsV2::{
+    Bernoulli, Exponential, Gaussian, Uniform, Unknown,
+};
```
```suggestion
use datafusion_physical_expr_common::stats_v2::StatisticsV2::{
    self, Bernoulli, Exponential, Gaussian, Uniform, Unknown,
};
```

berkaysynnada committed:
bug fix

berkaysynnada committed:
apply reverse logic in stats framework as well

berkaysynnada committed:
Merge branch 'feat/statistics_v2' of https://github.com/Fly-Style/datafusion into feat/statistics_v2

berkaysynnada committed:
Update cp_solver.rs

berkaysynnada committed:
revert data.parquet

ozankabak commented on datafusion/physical-expr/src/expressions/not.rs:
```diff
@@ -22,12 +22,16 @@ use std::fmt;
 use datafusion_expr::interval_arithmetic::Interval;
 use datafusion_expr::ColumnarValue;
+use datafusion_physical_expr_common::stats_v2::StatisticsV2;
+use datafusion_physical_expr_common::stats_v2::StatisticsV2::Bernoulli;
```
```suggestion
use datafusion_physical_expr_common::stats_v2::StatisticsV2::{self, Bernoulli};
```

ozankabak commented on datafusion/physical-expr/src/utils/mod.rs:
```diff
@@ -32,8 +32,9 @@ use datafusion_common::tree_node::{
 use datafusion_expr::Operator;
-
 use datafusion_physical_expr_common::sort_expr::LexOrdering;
+pub use guarantee::{Guarantee, LiteralGuarantee};
```
Since this is an export, it is better to place it right after the `mod` and `pub mod` statements.

ozankabak commented on datafusion/physical-expr/src/expressions/binary.rs:
```diff
@@ -32,14 +36,16 @@ use arrow::compute::{cast, ilike, like, nilike, nlike};
+use datafusion_physical_expr_common::stats_v2::StatisticsV2;
+use datafusion_physical_expr_common::stats_v2::StatisticsV2::{
+    Bernoulli, Gaussian, Uniform, Unknown,
+};
```
```suggestion
use datafusion_physical_expr_common::stats_v2::StatisticsV2::{
    self, Bernoulli, Gaussian, Uniform, Unknown,
};
```

ozankabak commented on datafusion/physical-expr/src/expressions/binary.rs:
```diff
@@ -404,7 +410,7 @@ impl PhysicalExpr for BinaryExpr {
             if interval.eq(&Interval::CERTAINLY_TRUE) {
                 // A certainly true logical conjunction can only derive from possibly
-                // true operands. Otherwise, we prove infeasibility.
+                // true operands. Otherwise, we prove infeasability.
```
```suggestion
                // true operands. Otherwise, we prove infeasibility.
```

ozankabak commented on datafusion/physical-expr/src/expressions/binary.rs:
```diff
@@ -444,7 +450,7 @@ impl PhysicalExpr for BinaryExpr {
             if interval.eq(&Interval::CERTAINLY_FALSE) {
                 // A certainly false logical conjunction can only derive from certainly
-                // false operands. Otherwise, we prove infeasibility.
+                // false operands. Otherwise, we prove infeasability.
```
```suggestion
                // false operands. Otherwise, we prove infeasibility.
```

ozankabak commented on datafusion/physical-expr/src/expressions/binary.rs:
```diff
@@ -4387,4 +4500,322 @@ mod tests {
+        }
+        Ok(())
+    }
+    //endregion evaluate_statistics test
```
```suggestion
```

ozankabak commented on datafusion/physical-expr/src/expressions/binary.rs:
```diff
@@ -4387,4 +4500,322 @@ mod tests {
     }
+    //endregion
+
+    //region evaluate_statistics and propagate_statistics test
```
```suggestion
```

ozankabak commented on datafusion/physical-expr/src/expressions/binary.rs:
```diff
@@ -841,6 +952,8 @@ mod tests {
     }
 
+    //region tests
+
```
```suggestion
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,795 @@
+    LN_TWO_LOCK.get_or_init(|| ScalarValue::Float64(Some(2_f64.ln())))
+}
+
+/// New, enhanced `Statistics` definition, represents five core definitions.
```
```suggestion
/// New, enhanced `Statistics` definition, represents five core statistical distributions. New variants will be added over time.
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,795 @@
+/// New, enhanced `Statistics` definition, represents five core definitions.
+#[derive(Clone, Debug, PartialEq)]
+pub enum StatisticsV2 {
+    Uniform {
```
```suggestion
    /// Uniform distribution, represented by its range. For a more in-depth discussion, see:
    ///
    /// https://en.wikipedia.org/wiki/Continuous_uniform_distribution
    Uniform {
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,795 @@
+    Uniform {
+        interval: Interval,
+    },
+    /// f(x, λ, offset) = (λe)^(-λ(x - offset)), if x >= offset
```
```suggestion
    /// Exponential distribution with an optional shift, whose PDF is as follows:
    /// f(x, λ, offset) = (λe)^(-λ(x - offset)), if x >= offset, 0 otherwise.
    /// For a more in-depth discussion, see:
    ///
    /// https://en.wikipedia.org/wiki/Exponential_distribution
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,795 @@
+        rate: ScalarValue,
+        offset: ScalarValue,
+    },
+    Gaussian {
```
```suggestion
    /// Gaussian (normal) distribution, represented by its mean and variance.
    /// For a more in-depth discussion, see:
    ///
    /// https://en.wikipedia.org/wiki/Normal_distribution
    Gaussian {
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,795 @@
+        mean: ScalarValue,
+        variance: ScalarValue,
+    },
+    /// `p` always has [`DataType::Float64`]
```
```suggestion
    /// Bernoulli distribution with success probability `p`, which always has the
    /// data type [`DataType::Float64`]. For a more in-depth discussion, see:
    ///
    /// https://en.wikipedia.org/wiki/Bernoulli_distribution
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,795 @@
+    Bernoulli {
+        p: ScalarValue,
+    },
+    Unknown {
```
```suggestion
    /// An unknown distribution, only containing some summary statistics.
    /// For a more in-depth discussion, see:
    ///
    /// https://en.wikipedia.org/wiki/Summary_statistics
    Unknown {
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,795 @@
+            );
+        }
+
+        let stat = Uniform { interval };
+        if stat.is_valid()? {
+            Ok(stat)
+        } else {
+            internal_err!("Tried to construct invalid Uniform statistic")
+        }
```
```suggestion
        Ok(Uniform { interval })
```
Seems like we don't check `Uniform` variants in the `is_valid` function. We have two options here:
1. Add a match arm to `is_valid` to ensure that neither the lower nor the upper bound of range is infinite. This would give us a strict, proper distribution.
2. Allow improper uniform distributions with unbounded endpoints. In this case, we can use the suggestion in this comment.

Let's go with option (2) for now, we will switch to (1) in the future if this proves to be a problem.

berkaysynnada replied:
https://github.com/synnada-ai/datafusion-upstream/pull/57#discussion_r1938308801

~How do we represent infinite bounds then?~

This is uniform, you are right

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,795 @@
+
+impl StatisticsV2 {
+    /// Constructs a new [`StatisticsV2`] with [`Uniform`] distribution from given [`Interval`],
+    /// and checks newly created statistic on validness.
```
```suggestion
    /// Constructs a new [`StatisticsV2`] with [`Uniform`] distribution from the given [`Interval`].
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,795 @@
+    }
+
+    /// Constructs a new [`StatisticsV2`] with [`Exponential`] distribution from given
+    /// rate and offset and checks newly created statistic on validness.
```
```suggestion
    /// Constructs a new [`StatisticsV2`] with [`Exponential`] distribution from the given
    /// rate/offset pair, and checks the newly created statistic for validity.
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,795 @@
+    }
+
+    /// Constructs a new [`StatisticsV2`] with [`Gaussian`] distribution from given
+    /// mean and variance and checks newly created statistic on validness.
```
```suggestion
    /// Constructs a new [`StatisticsV2`] with [`Gaussian`] distribution from the given
    /// mean/variance pair, and checks the newly created statistic for validity.
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,795 @@
+    }
+
+    /// Constructs a new [`StatisticsV2`] with [`Bernoulli`] distribution from given probability,
+    /// and checks newly created statistic on validness.
```
```suggestion
    /// Constructs a new [`StatisticsV2`] with [`Bernoulli`] distribution from the given probability,
    /// and checks the newly created statistic for validity.
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,795 @@
+
+    /// Constructs a new [`StatisticsV2`] with [`Unknown`] distribution from given
+    /// mean, median, variance, which are optional and range. Additionally, constructor
+    /// checks newly created statistic on validness.
```
```suggestion
    /// Constructs a new [`StatisticsV2`] with [`Unknown`] distribution from the given
    /// mean (optional), median (optional), variance (optional), and range values.
    /// Then, checks the newly created statistic for validity.
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,795 @@
+
+    /// Creates a new [`Unknown`] statistics instance with a given range.
+    /// It makes its best to infer mean, median and variance, if it is possible.
+    /// This builder is moved here due to original package visibility limitations.
```
```suggestion
    /// Constructs a new [`Unknown`] statistics instance from the given range.
    /// It makes heuristic estimates for mean, median and variance.
    /// This builder is moved here due to original package visibility limitations.
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,795 @@
+    /// This builder is moved here due to original package visibility limitations.
+    pub fn new_unknown_from_interval(range: &Interval) -> Result<StatisticsV2> {
+        // Note: to avoid code duplication for mean/median/variance computation, we wrap
+        // existing range in temporary uniform distribution and compute all these properties.
```
```suggestion
        // To avoid code duplication for mean/median/variance computations, we wrap the
        // given range in an ancillary uniform distribution and compute summary statistics from it.
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,795 @@
+    /// - For [`Unknown`],
+    ///   - if `mean` and/or `median` are defined, the `range` must contain their values
+    ///   - if `std_dev` is defined, it must be non-negative
+    fn is_valid(&self) -> Result<bool> {
```
Do we use `is_valid` other than in constructors? AFAICT we make sure to construct valid objects and maintain validity then on. If that is the case, it may be a good idea to move the code in the match arms of this function to their respective constructors and remove this function.

If the function finds use in tests, or other code (which I am missing right now), we should keep it. I'd like to hear your thoughts on this.

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,795 @@
+    /// - [`Exponential`] distribution mean is calculable by formula: 1/λ. λ must be non-negative.
+    /// - [`Gaussian`] distribution has it explicitly.
+    /// - [`Bernoulli`] mean is `p`.
+    /// - [`Unknown`] distribution _may_ have it explicitly.
```
```suggestion
    /// Extracts the mean value of the given statistic, depending on its distribution:
    /// - A [`Uniform`] distribution's interval determines its mean value, which is the
    ///   arithmetic average of the interval endpoints.
    /// - An [`Exponential`] distribution's mean is calculable by formula `offset + 1/λ`,
    ///   where λ is the (non-negative) rate.
    /// - A [`Gaussian`] distribution contains the mean explicitly.
    /// - A [`Bernoulli`] distribution's mean is equal to its success probability `p`.
    /// - An [`Unknown`] distribution _may_ have it explicitly, or this information may
    ///   be absent.
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,795 @@
+                agg.div(ScalarValue::Float64(Some(2.))).map(Some)
+            }
+            Exponential { rate, .. } => {
+                let one = ScalarValue::new_one(&rate.data_type())?;
```
This formula seems wrong, it doesn't account for the shift (offset). See my suggestions for the docstring.

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,795 @@
+    /// - [`Exponential`] distribution median is calculable by formula: ln2/λ. λ must be non-negative.
+    /// - [`Gaussian`] distribution median is equals to mean, which is present explicitly.
+    /// - [`Bernoulli`] median is `1`, of `p > 0.5`, `0` otherwise.
+    /// - [`Unknown`] distribution median _may_ be present explicitly.
```
```suggestion
    /// Extracts the median value of the given statistic, depending on its distribution:
    /// - A [`Uniform`] distribution's interval determines its median value, which is the
    ///   arithmetic average of the interval endpoints.
    /// - An [`Exponential`] distribution's median is calculable by the formula `offset + ln2/λ`,
    ///   where λ is the (non-negative) rate.
    /// - A [`Gaussian`] distribution's median is equal to its mean, which is specified explicitly.
    /// - A [`Bernoulli`] distribution's median is `1` if `p > 0.5` and `0` otherwise.
    /// - An [`Unknown`] distribution _may_ have it explicitly, or this information may
    ///   be absent.
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,795 @@
+                    .cast_to(&DataType::Float64)?;
+                agg.div(ScalarValue::Float64(Some(2.))).map(Some)
+            }
+            Exponential { rate, .. } => get_ln_two().div(rate).map(Some),
```
Similar to the mean, effect of the offset seems to be missing here.

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,795 @@
+    }
+
+    /// Returns a data type of the statistics, if present
+    /// explicitly, or able to determine implicitly.
```
```suggestion
    /// Returns the data type of the statistics.
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,795 @@
+    /// - [`Exponential`]'s variance is equal to mean value and calculable as 1/λ^2
+    /// - [`Gaussian`]'s variance is available explicitly
+    /// - [`Bernoulli`]'s variance is `p*(1-p)`
+    /// - [`Unknown`]'s distribution variance _may_ be present explicitly.
```
```suggestion
    /// Extracts the variance value of the given statistic, depending on its distribution:
    /// - A [`Uniform`] distribution's interval determines its variance value, which is calculable
    ///   by the formula `(upper - lower) ^ 2 / 12`.
    /// - An [`Exponential`] distribution's variance is calculable by the formula `1/(λ ^ 2)`,
    ///   where λ is the (non-negative) rate.
    /// - A [`Gaussian`] distribution's variance is specified explicitly.
    /// - A [`Bernoulli`] distribution's median is given by the formula `p * (1 - p)` where `p`
    ///   is the success probability.
    /// - An [`Unknown`] distribution _may_ have it explicitly, or this information may
    ///   be absent.
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,795 @@
+    /// - [`Bernoulli`]'s returns [`Interval::UNCERTAIN`], if p != 0 and p != 1.
+    ///   Otherwise, returns [`Interval::CERTAINLY_FALSE`] and [`Interval::CERTAINLY_TRUE`],
+    ///   respectfully.
+    /// - [`Unknown`]'s range is unbounded by default, but
```
```suggestion
    /// Extracts the range of the given statistic, depending on its distribution:
    /// - A [`Uniform`] distribution's range is simply its interval.
    /// - An [`Exponential`] distribution's range is `[offset, +∞)`.
    /// - A [`Gaussian`] distribution's range is unbounded.
    /// - A [`Bernoulli`] distribution's range is [`Interval::UNCERTAIN`], if `p` is neither `0` nor `1`.
    ///   Otherwise, it is [`Interval::CERTAINLY_FALSE`] and [`Interval::CERTAINLY_TRUE`], respectively.
    /// - An [`Unknown`] distribution is unbounded by default, but more information may be present.
```

ozankabak reviewed (commented):
The PR looks very good. I reviewed it carefully, everything except `stats_v2_graph.rs`. Let's address my comments and I will review the last remaining file in depth.

I only have one architectural concern: When we are making calculations to infer summary statistics, we automatically resort to `Float64`. But doing this doesn't make sense if we have a `Uniform` with certain data types; e.g. `Decimal`. Seems like we may want a simple coercion/promotion table (or logic) that maps distribution data types to summary statistic data types.

I am looking forward to hearing your thoughts on this.

ozankabak committed:
Apply suggestions from code review

ozankabak commented:
I committed the straightforward suggestions, there are only 4 inline comments remaining for you in addition to the main architectural question. We are very close to finish line. Great work.

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,815 @@
+    /// Uniform distribution, represented by its range. For a more in-depth discussion, see:
+    ///
+    /// https://en.wikipedia.org/wiki/Continuous_uniform_distribution
+    Uniform {
+        interval: Interval,
+    },
```
```suggestion
    Uniform { interval: Interval },
```

ozankabak committed:
Update datafusion/physical-expr-common/src/stats_v2.rs

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,813 @@
+    /// data type [`DataType::Float64`]. For a more in-depth discussion, see:
+    ///
+    /// https://en.wikipedia.org/wiki/Bernoulli_distribution
+    Bernoulli {
+        p: ScalarValue,
+    },
```
```suggestion
    Bernoulli { p: ScalarValue },
```

ozankabak committed:
Update datafusion/physical-expr-common/src/stats_v2.rs

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,811 @@
+pub enum StatisticsV2 {
+    /// Uniform distribution, represented by its range. For a more in-depth discussion, see:
+    ///
+    /// https://en.wikipedia.org/wiki/Continuous_uniform_distribution
```
```suggestion
    /// <https://en.wikipedia.org/wiki/Continuous_uniform_distribution>
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,811 @@
+    /// f(x, λ, offset) = (λe)^(-λ(x - offset)), if x >= offset, 0 otherwise.
+    /// For a more in-depth discussion, see:
+    ///
+    /// https://en.wikipedia.org/wiki/Exponential_distribution
```
```suggestion
    /// <https://en.wikipedia.org/wiki/Exponential_distribution>
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,811 @@
+    /// Gaussian (normal) distribution, represented by its mean and variance.
+    /// For a more in-depth discussion, see:
+    ///
+    /// https://en.wikipedia.org/wiki/Normal_distribution
```
```suggestion
    /// <https://en.wikipedia.org/wiki/Normal_distribution>
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,811 @@
+    /// Bernoulli distribution with success probability `p`, which always has the
+    /// data type [`DataType::Float64`]. For a more in-depth discussion, see:
+    ///
+    /// https://en.wikipedia.org/wiki/Bernoulli_distribution
```
```suggestion
    /// <https://en.wikipedia.org/wiki/Bernoulli_distribution>
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,811 @@
+    /// An unknown distribution, only containing some summary statistics.
+    /// For a more in-depth discussion, see:
+    ///
+    /// https://en.wikipedia.org/wiki/Summary_statistics
```
```suggestion
    /// <https://en.wikipedia.org/wiki/Summary_statistics>
```

ozankabak committed:
Apply suggestions from code review

Fix links

Fly-Style committed:
Merge branch 'apache_main' into feat/statistics_v2

Fly-Style committed:
Fix compilation issue

Fly-Style commented:
@ozankabak thanks for the great review, thanks so much! I enjoy it a lot!

>  Seems like we may want a simple coercion/promotion table (or logic) that maps distribution data types to summary statistic data types.

Totally agree with you -- it is the most optimal way to do such calculation.

Fly-Style committed:
Fix mean/median formula for exponential distribution

berkaysynnada committed:
casting + exp dir + remove opt's + is_valid refractor

berkaysynnada committed:
Update stats_v2_graph.rs

berkaysynnada committed:
remove inner mod

berkaysynnada committed:
last todo: bernoulli propagation

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,887 @@
+    Exponential {
+        rate: ScalarValue,
+        offset: ScalarValue,
+        /// `true` if the exponential distribution has a positive tail (extending towards positive x),
+        /// `false` if it has a negative tail (extending towards negative x).
```
```suggestion
        /// Indicates whether the exponential distribution has a positive tail; i.e. it extends
        /// towards positive infinity.
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,887 @@
+    pub fn new_uniform(interval: Interval) -> Result<Self> {
+        if interval.data_type().eq(&DataType::Boolean) {
+            return internal_err!(
+                "Construction of boolean Uniform statistic is prohibited.\
+             Create Bernoulli statistic instead."
```
```suggestion
                "Construction of a boolean `Uniform` statistic is prohibited.\
             Create a `Bernoulli` statistic instead."
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,887 @@
+/// New, enhanced `Statistics` definition, represents five core statistical distributions. New variants will be added over time.
+#[derive(Clone, Debug, PartialEq)]
+pub enum StatisticsV2 {
+    /// Uniform distribution, represented by its range. For a more in-depth discussion, see:
+    ///
+    /// <https://en.wikipedia.org/wiki/Continuous_uniform_distribution>
```
```suggestion
    /// Uniform distribution, represented by its range. If the given range extends towards infinity,
    /// the distribution will be improper -- which is OK. For a more in-depth discussion, see:
    ///
    /// <https://en.wikipedia.org/wiki/Continuous_uniform_distribution>
    /// <https://en.wikipedia.org/wiki/Prior_probability#Improper_priors>    
```

ozankabak committed:
Apply suggestions from code review

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,889 @@
+        offset: ScalarValue,
+        positive_tail: bool,
+    ) -> Result<Self> {
+        if offset.is_null() {
```
I guess this check is outside `is_valid_exponential` because we don't have an actual constraint on its value. Yet, we still check for non-nullity because we don't yet handle exponentials with unknown offsets (IIRC). Therefore, maybe it makes sense to check for it inside is_valid_exponential until we start handling exponentials with unknown offsets.

berkaysynnada replied:
Making this distinction makes sense?:
- If the check is for a mathematical reason, include it in is_valid_...()
- If it is because of smth we "cannot support yet" or "don't prefer supporting", keep it in new_...()

alternatively, we could combine all checks under is_valid's?

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,889 @@
+    pub fn new_uniform(interval: Interval) -> Result<Self> {
+        if interval.data_type().eq(&DataType::Boolean) {
+            return internal_err!(
+                "Construction of a boolean `Uniform` statistic is prohibited.\
+             Create a `Bernoulli` statistic instead."
```
```suggestion
                "Construction of a boolean `Uniform` statistic is prohibited, create a `Bernoulli` statistic instead."
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,889 @@
+    ) -> Result<Self> {
+        if range.data_type().eq(&DataType::Boolean) {
+            internal_err!(
+                "Usage of boolean range during Unknown statistic construction \
+            is prohibited. Create Bernoulli statistic instead."
```
```suggestion
                "Construction of a boolean `Unknown` statistic is prohibited, create a `Bernoulli` statistic instead."
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,889 @@
+                "Offset argument of the Exponential distribution must be non-null"
+            )
+        } else if !is_valid_exponential(&rate)? {
+            internal_err!("Tried to construct invalid Exponential statistic")
```
```suggestion
            internal_err!("Tried to construct an invalid Exponential statistic")
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,889 @@
+            if valid {
+                Ok(Gaussian { mean, variance })
+            } else {
+                internal_err!("Tried to construct invalid Gaussian statistic")
```
```suggestion
                internal_err!("Tried to construct an invalid Gaussian statistic")
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,889 @@
+            if valid {
+                Ok(Bernoulli { p })
+            } else {
+                internal_err!("Tried to construct invalid Bernoulli statistic")
```
```suggestion
                internal_err!("Tried to construct an invalid Bernoulli statistic")
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,889 @@
+            is prohibited. Create Bernoulli statistic instead."
+            )
+        } else if !is_valid_unknown(&mean, &median, &variance, &range)? {
+            internal_err!("Tried to construct invalid Unknown statistic")
```
```suggestion
            internal_err!("Tried to construct an invalid Unknown statistic")
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,889 @@
+        match &self {
+            Exponential { rate, .. } => is_valid_exponential(rate),
+            Gaussian { variance, .. } => is_valid_gaussian(variance),
+            Bernoulli { p } => is_valid_gaussian(p),
```
```suggestion
            Bernoulli { p } => is_valid_bernoulli(p),
```

ozankabak committed:
Apply suggestions from code review

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,887 @@
+    pub fn median(&self) -> Result<ScalarValue> {
+        match &self {
+            Uniform { interval, .. } => {
+                let agg = interval
+                    .lower()
+                    .add_checked(interval.upper())?
+                    .cast_to(&DataType::Float64)?;
+                agg.div(ScalarValue::Float64(Some(2.)))
```
```suggestion
                let dt = interval.data_type();
                interval
                    .lower()
                    .add_checked(interval.upper())?
                    .div(ScalarValue::from(2).cast_to(dt))
```
This seems to polymorphize this section, it may be better to do it this way until we implement proper type promotion.

berkaysynnada committed:
prop_stats in binary

berkaysynnada committed:
Update binary.rs

berkaysynnada committed:
rename intervals

berkaysynnada committed:
block explicit construction

berkaysynnada committed:
test updates

berkaysynnada committed:
Update binary.rs

berkaysynnada committed:
revert renaming

ozankabak commented on datafusion/physical-expr-common/src/physical_expr.rs:
```diff
@@ -114,7 +115,7 @@ pub trait PhysicalExpr: Send + Sync + Display + Debug + DynEq + DynHash {
     /// expression.
     ///
-    /// This is used to propagate constraints down through an expression tree.
+    /// This is used to propagate ranges down through an expression tree.
```
```suggestion
    /// This is used to propagate constraints down through an expression tree.
```

ozankabak commented on datafusion/physical-expr-common/src/physical_expr.rs:
```diff
@@ -125,7 +126,7 @@ pub trait PhysicalExpr: Send + Sync + Display + Debug + DynEq + DynHash {
     /// A `Vec` of new intervals for the children, in order.
     ///
-    /// If constraint propagation reveals an infeasibility for any child, returns
+    /// If range propagation reveals an infeasibility for any child, returns
```
```suggestion
    /// If constraint propagation reveals an infeasibility for any child, returns
```

berkaysynnada committed:
impl range methods as well

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,955 @@
+    SCALAR_VALUE_ONE_LOCK.get_or_init(|| ScalarValue::Float64(Some(1.)))
+}
+
+/// Returns a ln(2) as a [`ScalarValue`]
```
```suggestion
/// Returns the number `ln(2)` as a [`ScalarValue`].
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,955 @@
+    SCALAR_VALUE_ZERO_LOCK.get_or_init(|| ScalarValue::Float64(Some(0.)))
+}
+
+/// Returns a `1` as a [`ScalarValue`]
```
```suggestion
/// Returns a `1` as a [`ScalarValue`].
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,955 @@
+static SCALAR_VALUE_ONE_LOCK: OnceLock<ScalarValue> = OnceLock::new();
+static LN_TWO_LOCK: OnceLock<ScalarValue> = OnceLock::new();
+
+/// Returns a `0` as a [`ScalarValue`]
```
```suggestion
/// Returns a `0` as a [`ScalarValue`].
```

ozankabak committed:
Apply suggestions from code review

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,955 @@
+}
+
+impl StatisticsV2 {
+    /// Constructs a new [`StatisticsV2`] with [`Uniform`] distribution from the given [`Interval`].
```
```suggestion
    /// Constructs a new [`StatisticsV2`] instance with [`Uniform`] distribution from the given [`Interval`].
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,955 @@
+        UniformDistribution::new(interval).map(Uniform)
+    }
+
+    /// Constructs a new [`StatisticsV2`] with [`Exponential`] distribution from the given
```
```suggestion
    /// Constructs a new [`StatisticsV2`] instance with [`Exponential`] distribution from the given
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,955 @@
+        ExponentialDistribution::new(rate, offset, positive_tail).map(Exponential)
+    }
+
+    /// Constructs a new [`StatisticsV2`] with [`Gaussian`] distribution from the given
```
```suggestion
    /// Constructs a new [`StatisticsV2`] instance with [`Gaussian`] distribution from the given
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,955 @@
+        GaussianDistribution::new(mean, variance).map(Gaussian)
+    }
+
+    /// Constructs a new [`StatisticsV2`] with [`Bernoulli`] distribution from the given probability,
```
```suggestion
    /// Constructs a new [`StatisticsV2`] instance with [`Bernoulli`] distribution from the given probability,
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,955 @@
+        BernoulliDistribution::new(p).map(Bernoulli)
+    }
+
+    /// Constructs a new [`StatisticsV2`] with [`Unknown`] distribution from the given mean,
```
```suggestion
    /// Constructs a new [`StatisticsV2`] instance with [`Unknown`] distribution from the given mean,
```

ozankabak committed:
Apply suggestions from code review

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,955 @@
+///
+/// For a positive tail (when `positive_tail` is `true`):
+///
+/// `f(x; λ, offset) = λ exp(-λ (x - offset))    for x ≥ offset`
```
```suggestion
/// `f(x; λ, offset) = λ exp(-λ (x - offset)) for x ≥ offset, 0 otherwise`
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,955 @@
+///
+/// For a negative tail (when `positive_tail` is `false`):
+///
+/// `f(x; λ, offset) = λ exp(-λ (offset - x))    for x ≤ offset`
```
```suggestion
/// `f(x; λ, offset) = λ exp(-λ (offset - x)) for x ≤ offset, 0 otherwise.`
```

ozankabak commented on datafusion/physical-expr-common/src/stats_v2.rs:
```diff
@@ -0,0 +1,955 @@
+    }
+
+    pub fn target_type(args: &[&ScalarValue]) -> Result<DataType> {
+        let mut typed_args = args.iter().filter_map(|&arg| {
+            if arg == &ScalarValue::Null {
+                None
+            } else {
+                Some(arg.data_type())
+            }
+        });
+
+        typed_args
+            .next()
+            .map_or(Some(DataType::Null), |first| {
+                typed_args
+                    .try_fold(first, |target, arg| binary_numeric_coercion(&target, &arg))
+            })
+            .ok_or(DataFusionError::Internal(
+                "Statistics can only be evaluated for numeric types".to_string(),
+            ))
+    }
```
```suggestion
        let mut arg_types = args.iter().filter_map(|&arg| {
            if arg == &ScalarValue::Null {
                None
            } else {
                Some(arg.data_type())
            }
        });

        let Some(dt) = arg_types
            .next()
            .map_or(Some(DataType::Null), |first| {
                arg_types
                    .try_fold(first, |target, arg| binary_numeric_coercion(&target, &arg))
            }) else {
                return internal_err!("Statistics can only be evaluated for numeric types");
            };
            Ok(dt)
    }
```

ozankabak replied:
```suggestion
        let mut arg_types = args.iter().filter_map(|&arg| {
            if arg == &ScalarValue::Null {
                None
            } else {
                Some(arg.data_type())
            }
        });

        let Some(dt) = arg_types
            .next()
            .map_or(Some(DataType::Null), |first| {
                arg_types
                    .try_fold(first, |target, arg| binary_numeric_coercion(&target, &arg))
            }) else {
            return internal_err!("Statistics can only be evaluated for numeric types");
        };
        Ok(dt)
    }
```

ozankabak committed:
Update datafusion/physical-expr-common/src/stats_v2.rs

berkaysynnada committed:
Update stats_v2.rs

berkaysynnada committed:
fmt

berkaysynnada committed:
fix bernoulli or eval

ozankabak committed:
Review

berkaysynnada committed:
fmt

ozankabak committed:
Review Part 2

berkaysynnada committed:
not propagate

berkaysynnada committed:
clean-up

ozankabak committed:
Review Part 3

ozankabak committed:
Review Part 4

ozankabak committed:
Review Part 5

ozankabak committed:
Review Part 6

ozankabak committed:
Review Part 7

ozankabak committed:
Review Part 8

ozankabak committed:
Review Part 9

ozankabak committed:
Review Part 10

ozankabak committed:
Review Part 11

ozankabak committed:
Review Part 12

ozankabak committed:
Review Part 13

ozankabak committed:
Review Part 14

ozankabak committed:
Review Part 15 | Fix equality comparisons between uniform distributions

ozankabak committed:
Review Part 16 | Remove unnecessary temporary file

ozankabak committed:
Review Part 17 | Leave TODOs for real-valued summary statistics

ozankabak committed:
Review Part 18

ozankabak committed:
Review Part 19 | Fix variance calculations

ozankabak committed:
Review Part 20 | Fix range calculations

ozankabak committed:
Review Part 21

ozankabak committed:
Review Part 22

ozankabak committed:
Review Part 23

ozankabak committed:
Review Part 24 | Add default implementations for evaluate_statistics and propagate_statistics

ozankabak committed:
Review Part 25 | Improve docs, refactor statistics graph code

ozankabak committed:
Review Part 26

ozankabak committed:
Review Part 27

ozankabak committed:
Review Part 28 | Remove get_zero/get_one, simplify propagation in statistics graph

ozankabak committed:
Review Part 29

ozankabak committed:
Review Part 30 | Move statistics-combining functions to core module, polish tests

ozankabak committed:
Review Part 31

ozankabak committed:
Review Part 32 | Module reorganization

ozankabak committed:
Review Part 33

ozankabak reviewed (approved):
This is in good shape now - let's open the PR in the upstream repository so we can start the process of merging it upstream.

@Fly-Style -- test coverage is not great, let's increase it bit by bit as we collect feedback from the upstream community. Some TODOs to tackle during the process:
- Writing tests for the `combine_bernoullis` function.
- Writing tests for the `combine_gaussians` function.
- Enhancing the `create_bernoulli_from_comparison` function by implementing the TODOs I left therein.

Fly-Style committed:
Add tests for bernoulli and gaussians combination

Fly-Style commented:
@ozankabak tests were added here : https://github.com/synnada-ai/datafusion-upstream/pull/57/commits/cc99a9df3141de9047bdfa4424a7a8d70ca20297

Re: 
> Enhancing the create_bernoulli_from_comparison function by implementing the TODOs I left therein.

I think, it is better to pass it as a separate PR to not overload the PR, which is already massive.

Will open a PR to upstream soon.

Owen-CH-Leung committed:
Fix CI job test-datafusion-pyarrow (#14790)

* Change to ubuntu-latest to observe the error msg

* Use python 3.8.18

* Remove tag bullseye

* Revert to tag bullseye & remove setup python step

* Install pip

mbrobbel committed:
Use `doc_auto_cfg`, logo and favicon for docs.rs (#14746)

dependabot[bot] committed:
chore(deps): bump sqllogictest from 0.27.1 to 0.27.2 (#14785)

Bumps [sqllogictest](https://github.com/risinglightdb/sqllogictest-rs) from 0.27.1 to 0.27.2.
- [Release notes](https://github.com/risinglightdb/sqllogictest-rs/releases)
- [Changelog](https://github.com/risinglightdb/sqllogictest-rs/blob/main/CHANGELOG.md)
- [Commits](https://github.com/risinglightdb/sqllogictest-rs/compare/v0.27.1...v0.27.2)

---
updated-dependencies:
- dependency-name: sqllogictest
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

Kontinuation committed:
Add additional info about memory reservation to the doc of MemoryPool (#14789)

zhuqi-lucas committed:
feat: Improve datafusion-cli memory usage and considering reserve mem… (#14766)

* feat: Improve datafusion-cli memory usage and considering reserve memory for the result batches

* Address new comments

* Address new comments

* fix test

* fix test

* Address comments

* Fix doc

* Fix row count showing

* Fix fmt

* fix corner case

* remove unused code

2010YOUY01 committed:
Fix CI fail for extended test (by freeing up more disk space in CI runner) (#14745)

* Fix extended test

* feedback

* Update datafusion/core/tests/memory_limit/memory_limit_validation/sort_mem_validation.rs

Co-authored-by: Bruce Ritchie <bruce.ritchie@veeva.com>

* fix action version hash

---------

Co-authored-by: Bruce Ritchie <bruce.ritchie@veeva.com>

findepi committed:
chore: Benchmark deps cleanup (#14793)

* Synchronize criterion version across workspace

Use same version everywhere. This means update for some crates.

* Synchronize and update rand version across workspace

Use same version everywhere and update.

* Avoid linear search for character in bench helper

Get the char in O(1) instead of O(n).

findepi committed:
Fix test not to litter in repository (#14795)

Before the change, executing the test would create (and leave behind)
the `datafusion/execution/DOESNT_EXIST` folder inside source repository.

dependabot[bot] committed:
chore(deps): bump testcontainers from 0.23.2 to 0.23.3 (#14787)

Bumps [testcontainers](https://github.com/testcontainers/testcontainers-rs) from 0.23.2 to 0.23.3.
- [Release notes](https://github.com/testcontainers/testcontainers-rs/releases)
- [Changelog](https://github.com/testcontainers/testcontainers-rs/blob/main/CHANGELOG.md)
- [Commits](https://github.com/testcontainers/testcontainers-rs/compare/0.23.2...0.23.3)

---
updated-dependencies:
- dependency-name: testcontainers
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): bump serde from 1.0.217 to 1.0.218 (#14788)

Bumps [serde](https://github.com/serde-rs/serde) from 1.0.217 to 1.0.218.
- [Release notes](https://github.com/serde-rs/serde/releases)
- [Commits](https://github.com/serde-rs/serde/compare/v1.0.217...v1.0.218)

---
updated-dependencies:
- dependency-name: serde
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: jonahgao <jonahgao@msn.com>

logan-keede committed:
refactor: move `DataSource` to `datafusion-datasource` (#14671)

* build moves only tests+benches pending

* unstable

* some tests fixed

* Mock MemorySourceConfig and DataSource

* some cleanup

* test pass but Mock is not efficient

* temporary stable

* one struct, test pass, cleaning pending

* cleaning

* more cleaning

* clippy

* 🧹🧹*cleaning*🧹🧹

* adding re-export

* fix:cargo fmt

* fix: doctest

* fix: leftout doctest

* fix: circular dependency

* clean, rename, document, improve

mbrobbel committed:
Fix Clippy 1.85 warnings (#14800)

AdamGS committed:
Allow `FileSource`-specific repartitioning (#14754)

* FileSource specific repartitioning

* fix doc typo

* remove

* Avro doesn't support repartitioning

mbrobbel committed:
Bump MSRV to 1.82, toolchain to 1.85 (#14811)

* Bump MSRV to 1.82, toolchain to 1.85

* Fix some clippy warnings

* Fix more clippy warnings

timsaucer committed:
Chore/Add additional FFI unit tests (#14802)

* Add unit tests to FFI_ExecutionPlan

* Add unit tests for FFI table source

* Add round trip tests for volatility

* Add unit tests for FFI insert op

* Simplify string generation in unit test

Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

* Fix drop of borrowed value

---------

Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

timsaucer committed:
Chore/Add additional FFI unit tests (#14802)

* Add unit tests to FFI_ExecutionPlan

* Add unit tests for FFI table source

* Add round trip tests for volatility

* Add unit tests for FFI insert op

* Simplify string generation in unit test

Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

* Fix drop of borrowed value

---------

Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

jayzhan211 committed:
Remove CountWildcardRule in Analyzer and move the functionality in ExprPlanner, add `plan_aggregate` and `plan_window` to planner (#14689)

* count planner

* window

* update slt

* remove rule

* rm rule

* doc

* fix name

* fix name

* fix test

* tpch test

* fix avro

* rename

* switch to count(*)

* use count(*)

* rename

* doc

* rename window funciotn

* fmt

* rm print

* upd logic

* count null

jonahgao committed:
fix: normalize column names in table constraints (#14794)

* fix: normalize column names in table constraints

* newline

* Move slt

* restore ddl.slt

zhuqi-lucas committed:
fix: we are missing the unlimited case for bounded streaming when usi… (#14815)

* fix: we are missing the unlimited case for bounded streaming when using datafusion-cli

* Address comments

alamb committed:
Minor: comment in Cargo.toml about MSRV (#14809)

- I wasn't able to quickly find where the MSRV was defined when filing  https://github.com/apache/datafusion/issues/14808 so I would like to make it easier to find nex time

AdamGS committed:
Simplify `FileSource::create_file_opener`'s signature (#14798)

* simplify fn signature

* .

findepi committed:
Remove unused crate dependencies (#14827)

* Enable 'extended tests' on forks

Allow contributors to run extended tests workflow if they wish to, just
like they can run rust tests workflow on their forks, before opening a
PR to DataFusion. GitHub allows enabling/disabling workflows in the web
UI without needing to change workflow yaml file.

* Remove unused crate dependencies

Found by `cargo udeps`. Unfortunately there were false positives too.

* one on workspace level

ozankabak committed:
Incorporate community feedback

ozankabak committed:
Merge branch 'main' into feat/statistics_v2

ozankabak committed:
Fix merge issue

berkaysynnada commented:
https://github.com/apache/datafusion/pull/14699 is merged
